Source/Core/VideoCommon/TmemBase.cpp

```diff
@@ -0,0 +1,203 @@
+// Copyright 2019 Dolphin Emulator Project
+// Licensed under GPLv2+
+// Refer to the license.txt file included.
+
+#include <array>
+
+#include "VideoCommon/BPMemory.h"
+#include "VideoCommon/TmemBase.h"
+
+namespace TmemBase
+{
+// On TMEM configuration changed:
+// 1. invalidate stage.
+
+struct texture_unit_state
+{
+  enum State
+  {
+    INVALID, // Configuration has changed
+    VALID, // But not cached due to either being too big, or overlapping with another valid texture unit
+    CACHED,
+  };
+
+  TexImage1 even;
+  TexImage2 odd;
+  State state;
+
+  int even_base;
+  int even_size;
+  int odd_base;
+  int odd_size;
+};
+static std::array<texture_unit_state, 8> s_unit;
+
+void ConfigurationChanged(int bp_addr, int config)
+{
+  // Extract bits encoding texture unit
```

Comment:
Would you mind extending that comment with more info on what the bits in `bp_addr` mean (or, if it is easier: _why_ the shifting and masking is done)?

Reply:
Or better, use a struct to parse this BP command and pass on this struct instead of the raw u32.

Reply:
This change has been implemented in #10096